junit-vintage-engine/src/test/java/org/junit/vintage/engine/VintageTestEngineParallelismTests.java

```diff
@@ -0,0 +1,187 @@
+				"true").configurationParameter(DEFAULT_PARALLEL_MODE, "concurrent").configurationParameter(
+					PARALLEL_STRATEGY, "fixed").configurationParameter(PARALLELISM, "2").build();
+		TestEngine engine = new VintageTestEngine();
+		TestDescriptor descriptor = engine.discover(discoveryRequest, UniqueId.forEngine(engine.getId()));
+
+		CountingListener listener = new CountingListener(2);
+		ExecutionRequest executionRequest = new ExecutionRequest(descriptor, listener,
+			discoveryRequest.getConfigurationParameters());
+
+		// when
+		engine.execute(executionRequest);
+
+		// then
+		assertEquals(2, listener.getMaxConcurrentTests());
+		assertEquals(2, listener.getMaxConcurrentClasses());
+	}
+
+	/**
+	 * Verify that if parallel execution is disabled, tests will run sequentially.
+	 */
+	@Test
+	void verifyTestsRunSequentially() {
+		// given
+		LauncherDiscoveryRequest discoveryRequest = LauncherDiscoveryRequestBuilder.request().selectors(
+			selectClass(A.class), selectClass(B.class)).configurationParameter(PARALLEL_EXECUTION_ENABLED,
+				"false").build();
+		TestEngine engine = new VintageTestEngine();
+		TestDescriptor descriptor = engine.discover(discoveryRequest, UniqueId.forEngine(engine.getId()));
+
+		CountingListener listener = new CountingListener(1);
+		ExecutionRequest executionRequest = new ExecutionRequest(descriptor, listener,
+			discoveryRequest.getConfigurationParameters());
+
+		// when
+		engine.execute(executionRequest);
+
+		// then
+		assertEquals(1, listener.getMaxConcurrentTests());
+		assertEquals(1, listener.getMaxConcurrentClasses());
+	}
+
+	/**
+	 * {@link EngineExecutionListener} that exposes that maximum number of concurrent classes and tests.
+	 *
+	 * @since 5.8
+	 */
+	protected class CountingListener implements EngineExecutionListener {
+
+		private final IntSummaryStatistics classSummary = new IntSummaryStatistics();
+		private final IntSummaryStatistics testSummary = new IntSummaryStatistics();
+		private final LongAdder testAdder = new LongAdder();
+		private final LongAdder classAdder = new LongAdder();
+		private final CyclicBarrier barrier;
+
+		/**
+		 * @param methodConcurrency the number of test methods that may execute at the same time
+		 */
+		public CountingListener(int methodConcurrency) {
+			this.barrier = new CyclicBarrier(methodConcurrency);
+		}
+
+		/**
+		 * @return the largest number of tests that ran at the same time.
+		 */
+		public int getMaxConcurrentTests() {
+			return testSummary.getMax();
+		}
+
+		/**
+		 * @return the largest number of classes that ran at the same time.
+		 */
+		public int getMaxConcurrentClasses() {
+			return classSummary.getMax();
+		}
+
+		public void executionStarted(final TestDescriptor testDescriptor) {
+			String displayName = testDescriptor.getDisplayName();
+			if (testDescriptor.isTest()) {
+				incrementTests();
+			}
+			else if (displayName.equals("A") || displayName.equals("B")) {
```

Comment:
❓ Is this `if` necessary?

Reply:
Yes, because the outer container is "JUnit Vintage". I can change it to `else if( !testDescriptor.isRoot() ) {` if that's cleaner.

Reply:
I modified the `if`.